### ISSUE TITLE
crash in Geometry Optimization

### ISSUE THREAD

**Issue by dellelce** [psi4/psi4public#153]
I had a crash for a geometry optimization, this is the stack track:

Reading symbols from /app/miniconda/bin/psi4...(no debugging symbols found)...done.
[New LWP 7474]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `psi4'.
Program terminated with signal SIGABRT, Aborted.
#0  0x00006a91c31f18d7 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:55
55        return INLINE_SYSCALL (tgkill, 3, pid, selftid, sig);
(gdb) where
#0  0x00006a91c31f18d7 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:55
#1  0x00006a91c31f353a in __GI_abort () at abort.c:89
#2  0x00000000014c2e4f in opt::optking(psi::Options&) ()
#3  0x0000000000a4c2fc in py_psi_optking() ()
#4  0x0000000000a8993e in boost::python::objects::caller_py_function_impl<boost::python::detail::caller<int (*)(), boost::python::default_call_policies, boost::mpl::vector1<int> > >::operator()(_object*, _object*) ()
#5  0x000000000554b96e in boost::python::objects::function::call(_object*, _object*) const ()
#6  0x000000000554b6a5 in boost::detail::function::void_function_ref_invoker0<boost::python::objects::(anonymous namespace)::bind_return, void>::invoke(boost::detail::function::function_buffer&) ()
#7  0x000000000555699d in boost::python::detail::exception_handler::operator()(boost::function0<void> const&) const ()
#8  0x0000000000a6d6f5 in boost::detail::function::function_obj_invoker2<boost::_bi::bind_t<bool, boost::python::detail::translate_exception<psi::PsiException, void (*)(psi::PsiException const&)>, boost::_bi::list3<boost::arg<1>, boost::arg<2>, boost::_bi::value<void (*)(psi::PsiException const&)> > >, bool, boost::python::detail::exception_handler const&, boost::function0<void> const&>::invoke(boost::detail::function::function_buffer&, boost::python::detail::exception_handler const&, boost::function0<void> const&) ()
#9  0x0000000005556d54 in boost::python::handle_exception_impl(boost::function0<void>) ()
#10 0x000000000554d0fe in function_call ()
#11 0x00006a91c37de333 in PyObject_Call (func=0x9231260, arg=<optimized out>, kw=<optimized out>) at Objects/abstract.c:2529
#12 0x00006a91c388d4b2 in do_call (nk=<optimized out>, na=<optimized out>, pp_stack=0x712c327ea368, func=0x9231260) at Python/ceval.c:4253
#13 call_function (oparg=<optimized out>, pp_stack=0x712c327ea368) at Python/ceval.c:4058
#14 PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:2681
#15 0x00006a91c388fdbe in PyEval_EvalCodeEx (co=0x6a91b94db9b0, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=1, kws=0x92f85c8, kwcount=1, defs=0x0, defcount=0, closure=0x0) at Python/ceval.c:3267
#16 0x00006a91c388e3fa in fast_function (nk=<optimized out>, na=1, n=<optimized out>, pp_stack=0x712c327ea568, func=0x6a91b8cca488) at Python/ceval.c:4131
#17 call_function (oparg=<optimized out>, pp_stack=0x712c327ea568) at Python/ceval.c:4056
#18 PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:2681
#19 0x00006a91c388fdbe in PyEval_EvalCodeEx (co=0x6a91c45356b0, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=0, kws=0x0, kwcount=0, defs=0x0, defcount=0, closure=0x0) at Python/ceval.c:3267
#20 0x00006a91c388fed2 in PyEval_EvalCode (co=<optimized out>, globals=<optimized out>, locals=<optimized out>) at Python/ceval.c:669
#21 0x00006a91c38ae38c in run_mod (arena=0x93cee40, flags=<optimized out>, locals=0x6a91c4610d70, globals=0x6a91c4610d70, filename=0x6a91c38de42d "<string>", mod=<optimized out>) at Python/pythonrun.c:1371
#22 PyRun_StringFlags (
    str=0x92f8a04 "from psi4 import *\nfrom p4const import *\nfrom p4util import *\nfrom molutil import *\nfrom aliases import *\npsi4_io = psi4.IOManager.shared_object()\npsi4.efp_init()\npsi4_io.set_default_path(\"/app/scienc"..., 
    start=257, globals=0x6a91c4610d70, locals=0x6a91c4610d70, flags=<optimized out>) at Python/pythonrun.c:1334
#23 0x000000000555a8aa in boost::python::exec(boost::python::str, boost::python::api::object, boost::python::api::object) ()
#24 0x0000000000a67f25 in psi::Python::run(_IO_FILE*) ()
#25 0x00000000009937e0 in main ()
(gdb) 

the input file:

memory 2500 mb

molecule struct {
 symmetry c1
 no_reorient   #
 no_com   # whether to move the origin to the center of masses
 units angstrom
0 1   
      C                    0.008240    0.001327    0.000310
      C                    1.553694    0.001373    0.000036
      N                   -0.700795   -1.157242   -0.001319
      N                   -2.008585   -1.428031    0.002174
      N                   -0.699812    1.158389   -0.000832
      N                   -2.008322    1.432822    0.001677
      N                    2.267238    1.156444   -0.001011
      N                    3.565803    1.436891    0.001896
      N                    2.266013   -1.155613   -0.000583
      N                    3.563003   -1.435099    0.001415
      B                   -1.131831   -2.605160   -0.011643
      B                   -1.129410    2.608511   -0.010295
      B                    2.669048   -2.603366   -0.007508
      B                    2.671738    2.604971   -0.007555
}

set  {
  basis       6-31G*
  # Do use pure angular momentum basis functions?
 #  PUREAM false

 #  SCF_TYPE DIRECT  # An out-of-core, presorted algorithm using exact ERIs.

 #  MOLDEN_WRITE true

}

optimize('scf', dertype='energy')

psi4 is installed through miniconda:

[root@titan3 2]# find /app/ -name psi4 -type f
/app/miniconda/pkgs/psi4-0.2.54-py27_gf997694/bin/psi4
/app/miniconda/bin/psi4
[root@titan3 2]# /app/miniconda/bin/psi4 --version
0.2.54

**Comment by dellelce** [comment]
sorry can't edit the subject...

it should have been: "Crash in Geometry Optimization"...

It ran for about a day and then this.

**Comment by psi-rking** [comment]
A few preliminaries:
1. Does the optimization succeed if you don't use finite differences?
2. Does the same error occur if you don't freeze the com and prevent rotation?  I don't know if the automatically generated displacement coordinates make sense with these restrictions applied.
3. What is at the bottom of the output file, e.g., were any optimization steps taken?

**Comment by dellelce** [comment]
first 2 questions:
   Can try - this was my first psi4 optimization.

3:
"exception" related definitively to "new linear angles": 

```
        Energy change for the previous step:
                Projected    :        -0.0009583389
                Actual       :        -0.0004424593

        Performing BFGS update.
        Previous computed or guess Hessian on step 1.
        Steps to be used in Hessian update: 24 23
        Taking RFO optimization step.
        Going to follow RFO solution 1.
        Using RFO vector 1.
        Norm of target step-size    0.47532
        Projected energy change by RFO approximation:        -0.0041167176
        Newly linear bends that need to be incoporated into the internal coordinates:       1    3   11
            2    9   13
        The INTCO_EXCEPTion handler:
        New linear angles
        Dynamic level is 0.
        exc.g_really_quit() is 0.
    New linear angles
```

There are 1825 geometries in the output.dat file (it is about 22mb in size).

**Comment by loriab** [comment]
Here's same job with more "normal" input.
```
memory 2500 mb

molecule struct {
units angstrom
0 1
C 0.008240 0.001327 0.000310
C 1.553694 0.001373 0.000036
N -0.700795 -1.157242 -0.001319
N -2.008585 -1.428031 0.002174
N -0.699812 1.158389 -0.000832
N -2.008322 1.432822 0.001677
N 2.267238 1.156444 -0.001011
N 3.565803 1.436891 0.001896
N 2.266013 -1.155613 -0.000583
N 3.563003 -1.435099 0.001415
B -1.131831 -2.605160 -0.011643
B -1.129410 2.608511 -0.010295
B 2.669048 -2.603366 -0.007508
B 2.671738 2.604971 -0.007555
}

set basis 6-31G*
optimize('scf')
```
and output
```
  --------------------------------------------------------------------------------------------- ~
   Step     Total Energy     Delta E     MAX Force     RMS Force      MAX Disp      RMS Disp    ~
  --------------------------------------------------------------------------------------------- ~
    Convergence Criteria    1.00e-06 *    3.00e-04 *             o    1.20e-03 *             o  ~
  --------------------------------------------------------------------------------------------- ~
      1    -609.21574955   -6.09e+02      4.65e-02      1.39e-02 o    1.70e-01      5.38e-02 o  ~
      2    -609.24416188   -2.84e-02      1.95e-02      4.94e-03 o    3.67e-01      6.90e-02 o  ~
      3    -609.26373458   -1.96e-02      4.03e-02      1.34e-02 o    4.00e-01      1.12e-01 o  ~
      4    -609.29180689   -2.81e-02      4.13e-02      1.19e-02 o    3.28e-01      7.61e-02 o  ~
      5    -609.31034726   -1.85e-02      6.99e-02      1.39e-02 o    3.38e-01      6.27e-02 o  ~
      6    -609.33098015   -2.06e-02      4.29e-02      1.03e-02 o    4.77e-01      7.89e-02 o  ~
      7    -609.36524125   -3.43e-02      4.11e-02      7.82e-03 o    7.04e-01      1.05e-01 o  ~
      8    -609.41234162   -4.71e-02      4.58e-02      8.89e-03 o    5.68e-01      1.12e-01 o  ~
      9    -609.44676585   -3.44e-02      4.83e-02      1.17e-02 o    6.15e-01      9.91e-02 o  ~
     10    -609.49528129   -4.85e-02      4.44e-02      1.07e-02 o    4.17e-01      1.12e-01 o  ~
     11    -609.37833924    1.17e-01      1.22e-01      3.06e-02 o    2.26e-01      6.51e-02 o  ~
     12    -609.49194948   -1.14e-01      6.74e-02      1.50e-02 o    2.98e-01      6.15e-02 o  ~
     13    -609.50472895   -1.28e-02      4.41e-02      7.46e-03 o    6.15e-01      9.60e-02 o  ~
     14    -609.49299554    1.17e-02      5.21e-02      1.31e-02 o    4.25e-01      7.90e-02 o  ~
     15    -609.51141760   -1.84e-02      1.84e-02      4.83e-03 o    2.71e-01      9.38e-02 o  ~
     16    -609.46604175    4.54e-02      6.62e-02      1.90e-02 o    1.47e-01      5.38e-02 o  ~
     17    -609.50752379   -4.15e-02      3.93e-02      8.01e-03 o    1.44e-01      4.93e-02 o  ~
     18    -609.50243791    5.09e-03      4.01e-02      1.12e-02 o    2.42e-01      7.18e-02 o  ~
     19    -609.48911958    1.33e-02      7.26e-02      1.74e-02 o    1.09e-01      4.09e-02 o  ~
     20    -609.51528318   -2.62e-02      9.87e-03      2.56e-03 o    1.78e-01      5.73e-02 o  ~
     21    -609.51351330    1.77e-03      1.38e-02      3.38e-03 o    1.74e-01      4.56e-02 o  ~
     22    -609.51620753   -2.69e-03      3.27e-03      1.32e-03 o    5.00e-02      1.16e-02 o  ~
     23    -609.51663430   -4.27e-04      2.26e-03      7.58e-04 o    1.86e-01      2.83e-02 o  ~
     24    -609.51722373   -5.89e-04      2.81e-03      9.14e-04 o    2.33e-01      3.45e-02 o  ~
```

error is same as OP reports. Below shows starting (L) and final (R) geometries.

<img width="1357" alt="screen shot 2015-07-29 at 3 55 39 pm" src="https://cloud.githubusercontent.com/assets/2314730/8975665/110d0ab0-3650-11e5-9c6f-5358f4f84105.png">

**Comment by dellelce** [comment]
I am doing several more runs with several variations in parameters but the final result is the same, it crashes... the only difference is how many geometries there are before the crash.

**Comment by loriab** [comment]
Yes, I agree it looks messed up. Added the further info in last post for @psi-rking

**Comment by psi-rking** [comment]
It looks like an interesting optimization problem, and a memory error.  I
will take a detailed look at it when I get back from vacation next week.

**Comment by psi-rking** [comment]
optking was aborting prematurely.  I just pushed the fix to the problem.  For this case I found consecutive_backsteps = 2 [or more], as well as dynamic_level = 1 to now be helpful.  (Don't use both at once as that is somewhat redundant.)  Sometime I will type up some tips on problematic optimizations.  But now optking will at least move you substantially closer to the minimum.  Though one may always have to restart an optimization, these options will minimize the manual effort necessary.
